src/chainstate/stacks/boot/mod.rs

```diff
@@ -183,6 +185,72 @@ impl StacksChainState {
         .map(|value| value.expect_bool())
     }
 
+    /// Given a threshold and set of registered addresses, return a reward set where
+    ///   every entry address has stacked more than the threshold, and addresses
+    ///   are repeated floor(stacked_amt / threshold) times.
+    /// If an address appears in `addresses` multiple times, then the address's associated amounts
+    ///   are summed.
+    pub fn make_reward_set(
+        threshold: u128,
+        mut addresses: Vec<(StacksAddress, u128)>,
+    ) -> Vec<StacksAddress> {
+        let mut reward_set = vec![];
+        // the way that we sum addresses relies on sorting.
+        addresses.sort_by_key(|k| k.0.bytes.0);
+        while let Some((address, mut stacked_amt)) = addresses.pop() {
+            // peak at the next address in the set, and see if we need to sum
+            while addresses.last().map(|x| &x.0) == Some(&address) {
+                let (_, additional_amt) = addresses
+                    .pop()
+                    .expect("BUG: first() returned some, but pop() is none.");
+                stacked_amt = stacked_amt
+                    .checked_add(additional_amt)
+                    .expect("CORRUPTION: Stacker stacked > u128 max amount");
+            }
+            let slots_taken = u32::try_from(stacked_amt / threshold)
+                .expect("CORRUPTION: Stacker claimed > u32::max() reward slots");
+            info!(
+                "Slots taken by {} = {}, on stacked_amt = {}",
+                &address, slots_taken, stacked_amt
+            );
+            for _i in 0..slots_taken {
+                reward_set.push(address.clone());
+            }
+        }
+        reward_set
+    }
+
+    pub fn get_reward_threshold(
+        pox_settings: &PoxConstants,
+        addresses: &[(StacksAddress, u128)],
+        liquid_ustx: u128,
+    ) -> u128 {
+        let participation = addresses
+            .iter()
+            .fold(0, |agg, (_, stacked_amt)| agg + stacked_amt);
+
+        assert!(
+            participation <= liquid_ustx,
+            "CORRUPTION: More stacking participation than liquid STX"
+        );
+
+        let scale_by = cmp::max(participation, liquid_ustx / POX_MAXIMAL_SCALING as u128);
+
+        let reward_slots = pox_settings.reward_slots() as u128;
+        let threshold_precise = scale_by / reward_slots;
```

Comment:
Suggestion: remove `scale_by` in order to remove the double integer division.  The use of `cmp::max` may cause `threshold_precise` to be calculated as `(liquid_ustx / POX_MAXIMAL_SCALING) / reward_slots`, which risks losing precision.